sentry-core/src/main/java/io/sentry/core/Session.java

```diff
@@ -0,0 +1,288 @@
+    return sequence;
+  }
+
+  public void setSequence(Long sequence) {
+    this.sequence = sequence;
+  }
+
+  public Double getDuration() {
+    return duration;
+  }
+
+  public void setDuration(Double duration) {
+    this.duration = duration;
+  }
+
+  public Date getTimestamp() {
+    return timestamp;
+  }
+
+  public void setTimestamp(Date timestamp) {
+    this.timestamp = timestamp;
+  }
+
+  public void setUser(User user) {
+    this.user = user;
+  }
+
+  /**
+   * Starts a session and set its default values
+   *
+   * @param release the release
+   * @param environment the env
+   * @param user the user
+   * @param distinctId the distinct Id
+   */
+  public void start(
+      final String release, final String environment, final User user, final String distinctId) {
+    synchronized (sessionLock) {
+      init = true;
+      sequence = 0L;
+
+      if (sessionId == null) {
+        sessionId = UUID.randomUUID();
+      }
+
+      if (started == null) {
+        started = DateUtils.getCurrentDateTime();
+      }
+
+      if (release != null) {
+        this.release = release;
+      }
+
+      if (environment != null) {
+        this.environment = environment;
+      }
+
+      if (user != null) {
+        this.user = user;
+        if (this.user.getIpAddress() != null) {
+          ipAddress = this.user.getIpAddress();
+        }
+      }
+
+      if (distinctId != null) {
+        this.deviceId = distinctId;
+      }
+
+      if (status == null) {
+        status = State.Ok;
+      }
+    }
+  }
+
+  /** Updated the session status based on status and errorcount */
+  private void updateStatus() {
+    // at this state it might be Crashed already, so we don't check for it.
+    if (status == State.Ok && errorCount.get() > 0) {
+      status = State.Abnormal;
+    }
+
+    if (status == State.Ok) {
+      status = State.Exited;
+    }
+
+    // fallback if status is null
+    if (status == null) {
+      status = State.Ok;
+    }
+  }
+
+  /** Ends a session and update its values */
+  public void end() {
+    synchronized (sessionLock) {
+      init = null;
+      updateStatus();
+      timestamp = DateUtils.getCurrentDateTime();
+
+      // fallback if started is null
+      if (started == null) {
+        started = timestamp;
+      }
+
+      long diff =
+          Math.abs(timestamp.getTime() - started.getTime()); // do we need to subtract idle time?
+      duration = (double) diff;
+      sequence = System.currentTimeMillis();
+    }
+  }
+
+  /**
+   * Updates the current session and set its values
+   *
+   * @param status the status
+   * @param userAgent the userAgent
+   * @param addErrorsCount true if should increase error count or not
+   * @return if the session has been updated
+   */
+  public boolean update(final State status, final String userAgent, boolean addErrorsCount) {
+    synchronized (sessionLock) {
+      boolean sessionHasBeenUpdated = false;
+      if (status != null) {
```

Comment:
why is this racey?

Reply:
because in between the `status != null` and the line below modifying it, something can call `setStatus(null)`.
If we're synchronizing this, we need to make sure all access to this state is under the same lock object

Reply:
so all the other fields would be racey too, but the idea was to never expose Session object to a client anyway.
but moving this to a `ctor` would fix.